### PR TITLE
Support for W&B logger with VOC datasets

### PR DESCRIPTION
**Pull request by manangoel99**
This PR augments the current W&B logging infrastructure which worked with COCO style datasets to now also support logging and visualizing VOC datasets

### CODE REVIEW

**Comment by manangoel99** [comment]
Fixes #1412

**Comment by manangoel99** [comment]
Hi @FateScript I was wondering if it was possible to get a review on this?

**Comment by FateScript** [comment]
Hi @manangoel99 , thanks for reminding me of your pull request. There are so many PR in our repo and I will review them(including this) one by one tomorrow.
I'm busy with my work these days. Thanks for your patience!

File: yolox/exp/base_exp.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def __init__(self):
         self.seed = None
         self.output_dir = "./YOLOX_outputs"
         self.print_interval = 100
-        self.eval_interval = 10
```
Comment (FateScript) [inline yolox/exp/base_exp.py:24]: plz don't change default value here.

File: yolox/exp/yolox_base.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def __init__(self):
         self.print_interval = 10
         # eval period in epoch, for example,
         # if set to 1, model will be evaluate after every epoch.
-        self.eval_interval = 10
```
Comment (FateScript) [inline yolox/exp/yolox_base.py:94]: ditto

File: yolox/utils/logger.py [hunk 3]
```diff
@@ -13,6 +13,8 @@
 
 import torch
 
+from yolox.data.datasets import VOCDetection
```
Comment (FateScript) [inline yolox/utils/logger.py:16]: This might cause recurrent import issue, plz move it to `log_metric`

File: yolox/utils/logger.py [hunk 4]
```diff
@@ -276,17 +327,23 @@ def log_images(self, predictions):
...
+        if isinstance(self.val_dataset, VOCDetection):
+            predictions = self._convert_prediction_format(predictions)
 
         result_table = self.wandb.Table(columns=columns)
+
         for idx, val in table_ref.iterrows():
 
             avg_scores = defaultdict(int)
             num_occurrences = defaultdict(int)
+            if isinstance(val[0], list):
```
Comment (FateScript) [inline yolox/utils/logger.py:339]: what about
```
id = val[0]
if isinstance(id, list):
    id = id[0]
```

File: yolox/data/datasets/voc.py [hunk 5]
```diff
@@ -119,6 +119,10 @@ def __init__(
         self._annopath = os.path.join("%s", "Annotations", "%s.xml")
         self._imgpath = os.path.join("%s", "JPEGImages", "%s.jpg")
         self._classes = VOC_CLASSES
+        self.cats = [
+            {"id": idx, "name": val} for idx, val in enumerate(VOC_CLASSES)
+        ]
+        self.class_ids = [i for i in range(len(VOC_CLASSES))]
```
Comment (FateScript) [inline yolox/data/datasets/voc.py:125]: No need to iter here.
Try
```self.class_ids = list(range(len(VOC_CLASSES))```

File: yolox/utils/logger.py [hunk 6]
```diff
@@ -241,15 +244,63 @@ def _log_validation_set(self, val_dataset):
...
             self.val_artifact.add(self.val_table, "validation_images_table")
             self.run.use_artifact(self.val_artifact)
             self.val_artifact.wait()
 
+    def _convert_prediction_format(self, predictions):
+        image_wise_data = defaultdict(int)
+
+        for key, val in predictions.items():
+            img_id = key
+            # print(key, val)
```
Comment (FateScript) [inline yolox/utils/logger.py:265]: plz delete the comment.

File: yolox/utils/logger.py [hunk 7]
```diff
@@ -241,15 +244,63 @@ def _log_validation_set(self, val_dataset):
...
+            try:
+                bboxes = val[0]
+                cls = val[1]
+                scores = val[2]
+            except KeyError:
+                bboxes = val["bboxes"]
+                cls = val["categories"]
+                scores = val["scores"]
+
+            act_box = []
```
Comment (FateScript) [inline yolox/utils/logger.py:276]: what does `act` prefix mean here?

File: yolox/utils/logger.py [hunk 8]
```diff
@@ -241,15 +244,63 @@ def _log_validation_set(self, val_dataset):
...
+            act_scores = []
+            act_cls = []
+
+            if bboxes is not None:
+                for idx, box in enumerate(bboxes):
+                    box = bboxes[idx]
+                    cl = cls[idx]
```
Comment (FateScript) [inline yolox/utils/logger.py:283]: `cl` is a bad nameing for reading.

File: yolox/utils/logger.py [hunk 9]
```diff
@@ -241,15 +244,63 @@ def _log_validation_set(self, val_dataset):
...
+                bboxes = val["bboxes"]
+                cls = val["categories"]
+                scores = val["scores"]
+
+            act_box = []
+            act_scores = []
+            act_cls = []
+
+            if bboxes is not None:
+                for idx, box in enumerate(bboxes):
```
Comment (FateScript) [inline yolox/utils/logger.py:281]: Plz use zip here to make your code more clean.
```python
for box, class_id, score in zip(bboxes, cls, scores)
```

File: yolox/utils/logger.py [hunk 10]
```diff
@@ -241,15 +244,63 @@ def _log_validation_set(self, val_dataset):
...
 
+    def _convert_prediction_format(self, predictions):
+        image_wise_data = defaultdict(int)
+
+        for key, val in predictions.items():
+            img_id = key
+            # print(key, val)
+            # quit()
+            try:
+                bboxes = val[0]
```
Comment (FateScript) [inline yolox/utils/logger.py:268]: plz try
```
bboxes, cls, scores = val
```

File: yolox/utils/logger.py [hunk 11]
```diff
@@ -241,15 +244,63 @@ def _log_validation_set(self, val_dataset):
...
+                cls = val[1]
+                scores = val[2]
+            except KeyError:
+                bboxes = val["bboxes"]
```
Comment (FateScript) [inline yolox/utils/logger.py:272]: ditto,
plz try
```
bboxes, cls, scores = val["bboxes"], val["categries"], val["scores"]
```
It's more readable

**Review by FateScript — CHANGES_REQUESTED** [review]
PTAL @manangoel99

**Comment by FateScript** [comment]
@manangoel99  Any update?

**Comment by manangoel99** [comment]
Hi @FateScript ! Thank you so much for the comprehensive review! Apologies for the delay. I got caught up with some other work.
Will be resolving the comments today.

File: yolox/exp/base_exp.py (hunk 1 above)
Comment (manangoel99) [reply yolox/exp/base_exp.py:24]: Changed the values for testing locally.
Pushed it by mistake. I'll change it back

File: yolox/utils/logger.py (hunk 7 above)
Comment (manangoel99) [reply yolox/utils/logger.py:273]: This is for the actual bounding boxes i.e. list of bounding boxes after removing all the Nones.
Added a comment to explain this.

**Comment by manangoel99** [comment]
@FateScript I've added the suggested changes. Please have a look whenever you can.

**Comment by manangoel99** [comment]
I'm not sure why the test is failing
Screenshot 2022-10-31 at 5 46 51 PM

**Comment by FateScript** [comment]
@manangoel99 The workflow because of updating of importlib-meta. Reference stackoverflow [here](https://stackoverflow.com/questions/73929564/entrypoints-object-has-no-attribute-get-digital-ocean), could you plz also update the ci.yaml file under `.github/workflow` and using a fixed version of `importlib-metadata` ?

**Comment by manangoel99** [comment]
@FateScript I've set the version in ci.yml file

**Review by FateScript — APPROVED** [review]

File: yolox/exp/base_exp.py (hunk 1 above)
Comment (FateScript) [reply yolox/exp/base_exp.py:24]: @manangoel99 not changed back here?

File: yolox/exp/yolox_base.py (hunk 2 above)
Comment (FateScript) [reply yolox/exp/yolox_base.py:94]: @manangoel99 still not changed back.

**Review by FateScript — CHANGES_REQUESTED** [review]
@manangoel99 PTAL, some of your code still not change as I suggested.

**Comment by manangoel99** [comment]
@FateScript Apologies for this! I've fixed it now and pushed it.

**Review by FateScript — APPROVED** [review]

**Comment by FateScript** [comment]
LGTM, thanks for your contribution @manangoel99

**Pull request merged**